proposals/0012-introduce-reactNativeMeta.md

```diff
@@ -0,0 +1,148 @@
+---
+title: Introducing `reactNativeMeta` to `package.json`, for RN specific metadata
+author:
+  - Lorenzo Sciandra
+date: 25-01-2021
+---
+
+# RFC0012: Introducing `reactNativeMeta` to `package.json`, for RN specific metadata
+
+## Summary
+
+This RFC wants to introduce and formally discuss a new section for the package.json of a react-native project (app or library), called `reactNativeMeta`.
+
+This new section will allow developer to express certain characteristics of their code in a more formalized manner, that can easily used by tooling to act towards the code accordingly.
+
+**Note well**: this is but the first draft of the actual shape and fields. The goal of this draft is to collect feedback and help finalize the spec for this section.
+
+## Basic examples
+
+This is how this section would look like...
+
+...for a library:
+
+```json
+{
+    "name": "@rnx-kit/metro-serializer",
+    "version": "1.0.11",
+    ...,
+    "reactNativeMeta": {
+        "version": "2.3",
+        "type": "library",
+        "features": {
+            "fabric": "true",
+            "android": {
+                "turbomodules": "true",
+            },
+        },
+        "requirements": {
+            "react-native": ">= 0.71.0",
+            "expo": ">= 46.0.0",
```

Comment:
What extra information does this give over peerDependencies? Related to that, there may be cases where platforms are optional, vs not (e.g. a module may work on vs require a certain platform), which we may want to express.

Reply:
Edit: Ah, I can see this providing more specificity when allowed per-platform.

Reply:
it's also complicated by:
- npm / yarn differences (npm 7 auto install)
- Tree flattening algorithms for npm / yarn / workspaces

@brentvatne can provide more details

Reply:
peer dependencies brings in too many we don't want, like those listed by @kelset above. here's one example of how peer can break things pretty easily, and why I recommend not using peer dependencies for react-native in general:

let's say you are building a library that is built on some react-native and expo primitives that are relatively stable, like view and text from react-native, and so your library expresses a peer dependency as follows:

```
{
  "name": "react-native-peer-dependencies",
  "version": "1.0.0",
  "main": "index.js",
  "peerDependencies": {
    "react-native": ">=0.70.0",
    "expo": ">=47.0.0"
  }
}
```

now @kelset and the release crew release `react-native@0.71.0-rc.0` and you go to install it. oh no! that does not work, because peer dependencies do not work with any beta or rc versions. [npm has no intention to change this behavior](https://github.com/npm/npm/issues/8854). 

```
npm install react-native-peer-dependencies
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: example-app@1.0.0
npm ERR! Found: expo@47.0.0-beta.1
npm ERR! node_modules/expo
npm ERR!   expo@"47.0.0-beta.1" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer expo@">=47.0.0" from react-native-peer-dependencies@1.0.0
npm ERR! node_modules/react-native-peer-dependencies
npm ERR!   react-native-peer-dependencies@"*" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR! See /Users/brentvatne/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/brentvatne/.npm/_logs/2023-04-03T23_36_02_890Z-debug-0.log
```

Reply:
Thanks for pointing this out. I'm resolving this as `peerDependencies` got removed from this RFC as out of scope for now.